Implement and test user functions in the Qaul API

NoraCodes opened the pull request:
Just wanted to put this up to get feedback. It shouldn't be merged until I add some tests. See commit messages for more info.

NoraCodes committed:
Cargo fmt

NoraCodes committed:
Implement user_update for Qaul with new UserData

This consolidates the User struct into libqaul::users::User, rather than
having to import users::User as _User. It also seperates out the
UserData so that it's statically impossible to attempt to change the ID
of an existing user, and eliminate a lot of "id == id" checks.

identity::Identity is made public explicitly, since it snuck out in use
users::* and several other crates rely on it.

NoraCodes committed:
Add user_delete

Abstract the match for discrimination between Trusted and Default auth
into a method on UserAuth, require_trusted, which can be easily used
with the ? operator as seen in several of the user_ functions.

spacekookie commented on libqaul/http-api/src/auth.rs:
```diff
@@ -1,23 +1,17 @@
 use crate::QaulCore;
 use iron::{
-    BeforeMiddleware,
```
Could you maybe revert your `rustfmt` commit and only re-run it on files that you touched? That'll make the PR a bit more clear.

We should definitely pin some `fmt` task in CI to prevent accidentally merging things that aren't properly formatted (whoops :speak_no_evil:)

spacekookie commented on libqaul/src/api/models.rs:
```diff
@@ -28,6 +27,16 @@ pub enum UserAuth {
 
+impl UserAuth {
+    /// Returns an error if the UserAuth isn't Trusted.
+    pub fn require_trusted(self) -> QaulResult<(Identity, String)> {
```
That's a nice util function! :+1: 

Although maybe make it `pub(crate)` and shorten the name to something like `trusted` ? (just as an idea)

NoraCodes replied:
Yeah, good call.

spacekookie commented on libqaul/src/users/mod.rs:
```diff
@@ -4,14 +4,35 @@ use identity::Identity;
+    /// A users network ID
+    pub id: Identity,
+    /// The user's information, such as name, bio, etc.
+    pub data: UserData,
```
This is an interesting change. Initially I wanted to split the representation of a user `libqaul` internally and service API externally because internally we will probably also end up storing some secret stuff that we definitely don't want to accidentally leak outwards.

On the other had, we could probably just keep that state elsewhere and refer to it via the user id or something.

All in all, this feels more elegant, so :+1:

spacekookie reviewed (commented):
Cool! I left a few comments, although some of them are just "cool" :sweat_smile: 

Also note the thing about `rustfmt`, I'm gonna sit down and fix CI to include some checks with a pinned `fmt` or something

NoraCodes committed:
Refine naming and publicity of trusted()

NoraCodes committed:
Revert "Cargo fmt" (so as to undirty PR)

This reverts commit 06ff2a96265e24b1f8b7e58449d5cdf5cced94b7.

NoraCodes committed:
Cargo fmt

NoraCodes committed:
Implement contacts_add()

Add an identity() function for UserAuth which returns the identity no
matter what.

NoraCodes committed:
Add fallible resolver support for visn

This means changing the trait a bit; now it is generic over the return
type of the resolver as well, and the old infallible resolver is
implemented by simply making that equal to the system type parameter.

NoraCodes committed:
Derive common traits for QaulError

NoraCodes committed:
Add Clone derive for UserAuth

NoraCodes committed:
Builder-style attribute setters for UserData

NoraCodes committed:
Return correct authorization for new users

NoraCodes committed:
Basic test for creating/updating users

spacekookie commented on libqaul/src/api/mod.rs:
```diff
@@ -35,26 +37,48 @@ impl Qaul {
         users.insert(id.clone(), user);
 
-        Ok(UserAuth::Default(id))
+        Ok(UserAuth::Trusted(id, "fixme".into())) // TODO: Actually generate passphrase
```
I'd rather this line was a call to an `unimplemented!()` function in the `auth` module. We should let unimplemented things crash, rather than have "fixme" values :sweat_smile:

NoraCodes replied:
I chose not to do that because it would cause every test to autofail - if that's what we want, maybe I can make the unimplemented bit `#![cfg(not(test))]`? That feels wrong, somehow.

NoraCodes replied:
Oh, or, another option: we could have some way of injecting a user with the desired key material into the `Qaul` directly. I'll prototype that and see how well it works.

NoraCodes replied:
This is done in a714d91 and c689a18

spacekookie commented on libqaul/src/lib.rs:
```diff
@@ -2,19 +2,19 @@
 mod auth;
 mod crypto;
-mod users;
+pub mod users;
```
Actually I think maybe we keep the module private, then use fascading to expose only what people need (i.e. `User` and `UserData`. Considering there'll be more stuff in `users`, this feels more elegant and we can have a simple `api` prelude or something)

NoraCodes replied:
Definitely. I'll expose only those structs here.

spacekookie commented on libqaul/src/lib.rs:
```diff
@@ -2,19 +2,19 @@
     sync::{Arc, Mutex},
 };
-use users::User;
+use users::{User};
```
I guess this used to be more than one import? :D

NoraCodes replied:
Ah yeah, guess I forgot to run `rustfmt` for that last commit.

spacekookie reviewed (approved):
This is looking really really cool! I have a few nitpicky comments about style and export visibility but otherwise, it's great! :ok_hand: 

Happy to merge this after a few small changes.

NoraCodes committed:
Export libqual User and UserData explicitly

NoraCodes committed:
Move test user insertion to a dedicated method

user_create now panics due to no way to generate key material

NoraCodes committed:
Use user_inject() to set up tests

This fixes the test failure due to unimplemented!() in user_create

NoraCodes committed:
Let user_get work for all users

spacekookie commented on libqaul/src/lib.rs:
```diff
@@ -2,19 +2,20 @@
 pub use api::*;
 
-use identity::Identity;
+pub use identity::Identity;
```
I just realised this is the exact same change as #257, so let's just resolve the conversation over there by merging this PR and we can take another stab at the API visibilities later :sweat_smile:

spacekookie commented on visn/src/lib.rs:
```diff
@@ -57,71 +57,127 @@ use std::collections::VecDeque;
+/// # Types
+/// - `System`: the type of the system under test.
+/// - `Event`: the type of synthetic events.
+/// - `Return`: the type returned by the `resolve` function. Can be the same as `S`, or
```
Still referencing `S` here.

Btw, I'm not against using not-single letter type names here, but am curious why you wanted to spell them out. `S`, `E` and `R` wouldn't have been _too_ obscure, probably

NoraCodes replied:
I personally found them really confusing while doing refactorings but that might not impact the end user as much, I'm not sure.

NoraCodes replied:
To expand on that: `S`, `E`, `R` aren't to bad by themselves, but with `F` and `G` for function types and `P` for the concrete internal success type for `FallibleEngineImpl`, it gets confusing.

spacekookie reviewed (commented):
Just some minor comments and then we'll merge this. Also I guess closing #257 because of this

NoraCodes committed:
Clean up visn documentation and usage

spacekookie reviewed (approved):
Excellent, looks great now!
